jgrapht-core/src/test/java/org/jgrapht/alg/shortestpath/YenShortestPathIteratorTest.java

```diff
@@ -0,0 +1,329 @@
+                                   Integer source, Integer target) {
+        int amountOfPathsToIterate = 10;
+        YenShortestPathIterator<Integer, DefaultWeightedEdge> it
+                = new YenShortestPathIterator<>(graph, source, target);
+        GraphPath<Integer, DefaultWeightedEdge> path;
+        double previousPathWeight = 0.0;
+        Set<GraphPath<Integer, DefaultWeightedEdge>> paths = new HashSet<>();
+        int i = 0;
+        for (; i < amountOfPathsToIterate && it.hasNext(); i++) {
+            path = it.next();
+            paths.add(path);
+            assertCorrectPath(graph, path, source, target);
+            assertTrue(previousPathWeight <= path.getWeight());
+            previousPathWeight = path.getWeight();
+        }
+        assertEquals(i, paths.size());
+    }
+
+    private void performAssertion(Graph<Integer, DefaultWeightedEdge> graph,
+                                  YenShortestPathIterator<Integer, DefaultWeightedEdge> it,
+                                  Integer source, Integer target, double expectedWeight, boolean hasNext) {
+        GraphPath<Integer, DefaultWeightedEdge> path = it.next();
+        assertEquals(expectedWeight, path.getWeight(), 1e-9);
+        assertCorrectPath(graph, path, source, target);
+        assertLooplessPath(path);
+        assertEquals(it.hasNext(), hasNext);
+    }
+
+    private void assertCorrectPath(Graph<Integer, DefaultWeightedEdge> graph,
+                                   GraphPath<Integer, DefaultWeightedEdge> path,
+                                   Integer source, Integer target) {
+        List<DefaultWeightedEdge> edgeList = path.getEdgeList();
+
+        double expectedWeight = path.getWeight();
+        double actualWeight = edgeList.stream().mapToDouble(graph::getEdgeWeight).sum();
+        assertEquals(expectedWeight, actualWeight, 1e-9);
+
+        if (edgeList.size() == 0) {
+            assertEquals(source, target);
+        } else {
+            assertEquals(graph.getEdgeSource(edgeList.get(0)), source);
+            assertEquals(graph.getEdgeTarget(edgeList.get(edgeList.size() - 1)), target);
+
+            if (edgeList.size() >= 2) {
+                Iterator<DefaultWeightedEdge> it = edgeList.iterator();
+                DefaultWeightedEdge curr = it.next();
+                DefaultWeightedEdge next;
+                while (it.hasNext()) {
+                    next = it.next();
+                    assertEquals(graph.getEdgeTarget(curr), graph.getEdgeSource(next));
+                    curr = next;
+                }
+            }
+        }
+    }
+
+    private void assertLooplessPath(GraphPath<Integer, DefaultWeightedEdge> path) {
+        Set<Integer> uniqueVertices = new HashSet<>(path.getVertexList());
+        assertEquals(path.getVertexList().size(), uniqueVertices.size());
+    }
+
+    private void getRandomGraph(Graph<Integer, DefaultWeightedEdge> graph, int n, double p) {
```

Comment:
* This graph is constructed randomly. If I remember it correctly, we don't use random tests in production code. @jsichi @jkinable @d-michail is this correct?
* Let's add some small Javadoc to other private methods to explain briefly what they do

Reply:
Correct, even when using a random graph generator, our tests always specify a seed to make it deterministic.

Reply:
I think in a couple of cases we have both (a) a few tests with a fixed seed and (b)  a few with time-based seed.